### PR TITLE
Typing for Lightning Package

### LINKED ISSUE #5023: fix typing in PL codebase - multiple PRs
## 🚀 Feature

add/fix typing in all PL codebase, this will yield in multiple PR as we prefer to split the work into smaller peace

### Additional context

in #5021  we introduced a new ignore list so after it is merged, take out a single ignored item and fix typing for that part 
> each item + its fix = one PR

### Sections

- [ ] `pytorch_lightning.callbacks.*` #7035
- [ ] `pytorch_lightning.core.*`  #7035
- [ ] `pytorch_lightning.accelerators.*`  #7035
- [ ] `pytorch_lightning.loggers.*` #7035
- [ ] `pytorch_lightning.logging.*` #7035
- [ ] `pytorch_lightning.metrics.*` <---| @hassiahk
- [ ] `pytorch_lightning.overrides.*`
- [ ] `pytorch_lightning.profiler.*`
- [ ] `pytorch_lightning.pt_overrides.*`
- [ ] `pytorch_lightning.plugins.*`  #7022
- [ ] `pytorch_lightning.root_module.*`
- [ ] `pytorch_lightning.trainer.*`
- [ ] `pytorch_lightning.distributed.*`
- [ ] `pytorch_lightning.tuner.*`
- [ ] `pytorch_lightning.utilities.*`
- [ ] `pl_examples.* by` <---| @janhenriklambrechts
- [ ] `benchmarks.*`
- [ ] `tests.*`

_This sections can be split even to smaller, most likely still stay per package/folder_

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

This should enable mypy compatible typing throughout the lightning package. At some point, I had to ignore the type checks from mypy due to known issues or no good workarounds.

Resolves #5023

### Completed Packages

- [x] lightning.accelerators
- [x] lightning.callbacks
- [x] lightning.core
- [x] lightning.distributed
- [x] lightning.loggers
- [ ] lightning.metrics
- [ ] lightning.overrides
- [ ] lightning.plugins
- [ ] lightning.profiler
- [ ] lightning.trainer
- [ ] lightning.tuner
- [ ] lightning.utilities

Edit: The remaining parts will come in separate PRs to prevent this PR from becoming even larger

Open for discussion: Should we also type the tests? I don't think we have to!

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@justusschock seems you are shooting the types very fast a file per minute, are you using some automated tool for it?

File: pytorch_lightning/__about__.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import time
 
 _this_year = time.strftime("%Y")
-__version__ = '1.3.0rc1'
+__version__ = "20210414"
```
Comment (Borda) [inline pytorch_lightning/__about__.py:4]:
```suggestion
__version__ = "20210414"
```
```suggestion
__version__ = "1.3.0rc1"
```

**Comment by justusschock** [comment]
@Borda no, that's just me :D But actually that wasn't that fast. I started 2 days ago, just didn't push it all

File: pytorch_lightning/__about__.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/__about__.py]: Yes, sorry for that, that happens If I install it. Sometimes I accidently commit it.

**Comment by Borda** [comment]
cool, it seems to be very large, could have each sub-package as separate PR, it will be easier to check/merge and lower the change of collision in meantime...

**Comment by carmocca** [comment]
@justusschock I had #7022 already for accelerator and plugins

**Comment by Borda** [comment]
@justusschock @awaelchli @carmocca pls sync on #5023 what you are working on :]

**Comment by justusschock** [comment]
@carmocca Accelerators was easy, basically a noop :) But good to know that i don't need to do plugins :)

**Comment by Borda** [comment]
@justusschock does this has a successor PR or how this was closed? :rabbit:

**Pull request closed without merging**